crates/electrum/src/v2.rs

```diff
@@ -0,0 +1,507 @@
+    fn default() -> Self {
+        Self {
+            graph_update: Default::default(),
+            chain_update: Default::default(),
+            keychain_update: Default::default(),
+        }
+    }
+}
+
+impl<'a, K, A: Anchor> ElectrumUpdate<K, A> {
+    pub fn missing_full_txs<A2>(
+        &'a self,
+        graph: &'a TxGraph<A2>,
+    ) -> impl Iterator<Item = &'a Txid> + 'a {
+        self.graph_update
+            .keys()
+            .filter(move |&&txid| graph.as_ref().get_tx(txid).is_none())
+    }
+
+    pub fn finalize<T>(self, seen_at: Option<u64>, new_txs: T) -> LocalUpdate<K, A>
+    where
+        T: IntoIterator<Item = Transaction>,
+    {
+        let mut graph_update = TxGraph::<A>::new(new_txs);
+        for (txid, anchors) in self.graph_update {
+            if let Some(seen_at) = seen_at {
+                let _ = graph_update.insert_seen_at(txid, seen_at);
+            }
+            for anchor in anchors {
+                let _ = graph_update.insert_anchor(txid, anchor);
+            }
+        }
+        dbg!(graph_update.full_txs().count());
+        LocalUpdate {
+            keychain: self.keychain_update,
+            graph: graph_update,
+            chain: self.chain_update,
+        }
+    }
+}
+
+impl<K> ElectrumUpdate<K, ConfirmationHeightAnchor> {
+    pub fn finalize_as_confirmation_time<T>(
+        self,
+        client: &Client,
+        seen_at: Option<u64>,
+        new_txs: T,
+    ) -> Result<LocalUpdate<K, ConfirmationTimeAnchor>, Error>
+    where
+        T: IntoIterator<Item = Transaction>,
+    {
+        let update = self.finalize(seen_at, new_txs);
+        let update_tip = update.chain.tip().expect("must have tip");
+
+        let relevant_heights = {
+            let mut visited_heights = HashSet::new();
+            update
+                .graph
+                .all_anchors()
+                .iter()
+                .map(|(a, _)| a.confirmation_height_upper_bound())
+                .filter(move |&h| visited_heights.insert(h))
+                .collect::<Vec<_>>()
+        };
+
+        let height_to_time = relevant_heights
+            .clone()
+            .into_iter()
+            .zip(
+                client
+                    .batch_block_header(relevant_heights)?
+                    .into_iter()
+                    .map(|bh| bh.time as u64),
+            )
+            .collect::<HashMap<u32, u64>>();
+
+        if update_tip.hash != client.block_header(update_tip.height as _)?.block_hash() {
+            // [TODO] We should alter the logic so we won't have to return an error. This is to
+            // [TODO] ensure obtained block times are "anchored" to our tip. If we exclude this, it
+            // [TODO] should be "safe" as well. Tx confirmation times would just slightly vary.
+            return Err(Error::Message(format!(
+                "tip changed during update: update_tip={:?}",
+                update_tip
+            )));
+        }
```

Comment:
I think we should just note that the confirmation time might not be precisely correct if there was a re-org. Throwing an error here is worse than returning slightly off times.

I think electrum really intends for you to use the merkle proof API. Best solution is to do that and attach confirmation time of each block to `LocalChain`.

Reply:
I think so too, but probably for a separate PR?

Reply:
Issue created: https://github.com/bitcoindevkit/bdk/issues/980